### ISSUE TITLE
Using your own data on mnist -  "unknown backprop operator type"

### ISSUE THREAD

**Issue by CuthbertJungle** [leonardvandriel/caffe2_cpp_tutorial#50]
Hi, thanks for writing these tutorials c++ help is very sparse for caffe2

I'm looking to move onto using my own data for the mnist example but as AddInput uses a DB this is
obscuring how exactly the "data" and "label" elements are created. (where "data" and "label" are my input and output training files)

i have the data blobs correct, i think. 
auto tensor = workspace.CreateBlob("data")->GetMutable<TensorCPU>();
auto tensor = workspace.CreateBlob("label")->GetMutable<TensorCPU>();
but these relate to the workspace. 

if i define "data" as in input to the network

    model.predict.AddInput("data");
    model.AddFcOps("data", "fc1", 156, 100, test);
    model.predict.AddDataOp("data", "fc1");
    model.AddFcOps("fc1", "relu", 100, 50, test);
    model.predict.AddReluOp("relu", "relu");
    model.AddFcOps("fc1", "relu", 50, 100, test);
    model.predict.AddReluOp("relu", "relu");
    model.AddFcOps("fc1", "relu", 100, 50, test);
    model.predict.AddReluOp("relu", "relu");
    model.AddFcOps("relu", "pred", 50, 8, test);
    model.predict.AddSoftmaxOp("pred", "softmax");

i get a run time error of :        
terminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at net_gradient.cc:211] . unknown backprop operator type: data 

any help would be greatly appreciated :-)

**Comment by leonardvandriel** [comment]
The default way to allow the model to access data is by providing a data source, like a database reader with `CreateDB` and `TensorProtosDBInput`. You can do this manually, by setting the `data` and `label` tensor before every `RunNet(..)` of the model. Take a look at `pretrained.cc` for an example:

    auto input = workspace.CreateBlob("data")->GetMutable<TensorCPU>();
    input->ResizeLike(tensor);
    input->ShareData(tensor);
    CAFFE_ENFORCE(workspace.RunNetOnce(predict_net));

Is that helpful?
